src/Hint/Extensions.hs

```diff
@@ -188,31 +190,41 @@ extensionsHint _ x =
         sl
         (comment (mkLangExts sl exts))
         (Just newPragma)
-        ( [RequiresExtension $ prettyExtension gone | x <- before \\ after, gone <- Map.findWithDefault [] x disappear] ++
-            [ Note $ "Extension " ++ prettyExtension x ++ " is " ++ reason x
+        ( [RequiresExtension (show gone) | x <- before \\ after, gone <- Map.findWithDefault [] x disappear] ++
+            [ Note $ "Extension " ++ show x ++ " is " ++ reason x
             | x <- explainedRemovals])
         [ModifyComment (toSS' (mkLangExts sl exts)) newPragma]
     | (L sl _,  exts) <- langExts $ pragmas (ghcAnnotations x)
-    , let before = map parseExtension exts
+    , let before = map lookupExt (filterEnabled exts)
     , let after = filter (`Set.member` keep) before
     , before /= after
     , let explainedRemovals
             | null after && not (any (`Map.member` implied) before) = []
             | otherwise = before \\ after
     , let newPragma =
-            if null after then "" else comment (mkLangExts sl $ map prettyExtension after)
+            if null after then "" else comment (mkLangExts sl $ map show after)
     ]
   where
+    filterEnabled :: [String] -> [String]
+    filterEnabled  = filter (not . isPrefixOf "No")
+
+    lookupExt :: String -> Extension
+    lookupExt s =
+      case find (\(FlagSpec n _ _ _) -> n == s) xFlags of
+        Just f -> flagSpecFlag f
+        Nothing -> error ("IMPOSSIBLE: Unrecognized language extension '" ++ s ++ "'")
```

Comment:
Is this branch literally impossible to hit? How so?

Reply:
If it has been compiled it has been given to `parsePragmasIntoDynFlags` which has given it to `getOptions` which does a check that the given extension is valid. If it's not valid, a parse error will result. Therefore we are assured that the extensions listed in the pragmas of the module are kosher.

Reply:
Do we have a test that an invalid pragma leads to a parse error "hint"?

Reply:
I haven't written anything however, if I give it a module
```
{-# LANGUAGE Foo #-}
```
The result is early termination with `Error: Parse error`.

Reply:
It will have gone down this case in `parseModuleEx`:
```
          Left msg -> do
            -- Parsing GHC flags from dynamic pragmas in the source
            -- has failed. When this happens, it's reported by
            -- exception. It's impossible or at least fiddly getting a
            -- location so we skip that for now. Synthesize a parse
            -- error.
            let loc = GHC.mkSrcLoc (mkFastString file) (1 :: Int) (1 :: Int)
            pure $ Left (ParseError (GHC.mkSrcSpan loc loc) msg "")
```

Reply:
We can't get to `extensionHint` without a `ParseModuleEx {}` and if we have one of those, we know the extensions expressed by the pragmas have been validated.

Reply:
Cool. Can we have slightly more comment on this Impossible case then. And can we have a test in this module with `{-# LANGUAGE InvalidExtension #-} -- @ERROR ParseError` (or however we express it) so the test and the code relying on it are nearby.

Reply:
(Assuming that's fairly easy to do, don't sweat if it turns out to be non-obvious!)

Reply:
It would seem the message originates here:
```
-- | Return either an idea (a parse error) or the module. In IO because might call the C pre processor.
parseModuleApply :: ParseFlags -> [Setting] -> FilePath -> Maybe String -> IO (Either Idea ModuleEx)
parseModuleApply flags s file src = do
    res <- parseModuleEx (parseFlagsAddFixities [x | Infix x <- s] flags) file src
    case res of
      Right r -> pure $ Right r
      Left (ParseError sl msg ctxt) ->
            pure $ Left $ classify [x | SettingClassify x <- s] $ rawIdeaN Error "Parse error" sl ctxt Nothing []
```

Reply:
Done and done.